README.md
## Welcome to the Aegeus IPFS system

Here we proudly bring you a preview of the Aegeus user interface.

![preview](docs/src/markdown/trail/img/bob-list-03-small.png)

A full walk through of the demo is here: [QmYd7FRVWogUQqwDUFhwrvbNMohfips5b4y5X38Ys6Khzr](https://ipfs.io/ipfs/QmYd7FRVWogUQqwDUFhwrvbNMohfips5b4y5X38Ys6Khzr/trail)

### Installing Docker

Currently, this installation requires a working [Docker](https://www.docker.com/community-edition#/download) environment.

To verify that your Docker env is setup properly, you can list the running containers like this ...

    $ docker ps
    CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES

### The Aegeus images

In total there are four Docker images to make up the complete system.

1. [aegeus/aegeusd](https://hub.docker.com/r/aegeus/aegeusd)
2. [aegeus/aegeus-ipfs](https://hub.docker.com/r/aegeus/aegeus-ipfs)
3. [aegeus/aegeus-jaxrs](https://hub.docker.com/r/aegeus/aegeus-jaxrs)
4. [aegeus/aegeus-webui](https://hub.docker.com/r/aegeus/aegeus-webui)

What follows is an installation guide for all four containers. However, if you already have IPFS and AEG running locally, you will not need to run these in Docker again.
For a mixed setup with already running IPFS & AEG service and newly hosted Docker services go [here](docs/src/markdown/setup/Setup-Mixed-Docker.md).

### Quickstart

Here is the quickstart to get the whole system running in no time ...

    export GATEWAYIP=[YOUR_PUBLIC_IP]

    docker run --detach --name aegd -p 29328:29328 --memory=300m --memory-swap=2g aegeus/aegeusd
    docker run --detach --name aeg-ipfs -p 4001:4001 -p 8080:8080 -e GATEWAYIP=$GATEWAYIP --memory=300m --memory-swap=2g aegeus/aegeus-ipfs
    docker run --detach --name aeg-jaxrs --link aegd:blockchain --link aeg-ipfs:ipfs --memory=100m --memory-swap=2g aegeus/aegeus-jaxrs
    docker run --detach --name aeg-webui -p 8082:8082 --link aegd:blockchain --link aeg-ipfs:ipfs --link aeg-jaxrs:jaxrs --memory=100m --memory-swap=2g --env NESSUS_WEBUI_LABEL=Bob aegeus/aegeus-webui

You should now be able to access the WebUI at: [http://127.0.0.1:8082/portal](http://127.0.0.1:8082/portal)

### Updating the installation

Remove all running containers

    docker rm -f `docker ps -aq`

Pull the latest image versions

    docker pull aegeus/aegeusd
    docker pull aegeus/aegeus-ipfs
    docker pull aegeus/aegeus-jaxrs
    docker pull aegeus/aegeus-webui

Then, start again by running these containers.

### Building this project

You can use the standard maven build process, like this

    mvn clean install

However, running the tests will require to have an IPFS and an AEG wallet instance running.
Please follow the instructions for the [mixed setup](docs/src/markdown/setup/Setup-Mixed-Docker.md) to get this going.

Merci


docs/Changelog.md
### Changelog

#### Aegeus-UI 1.0.0 Beta1

**Features**

* [#51][51] Display block count in UI
* [#64][64] Go straight to file selection if there is only one address when clicking the Files link

For details see [1.0.0 Beta1 features](https://github.com/AegeusCoin/aegeus-ui/issues?q=milestone%3A%221.0.0+Beta1%22+label%3Afeature)

**Tasks**

* [#23][23] Update to aegeusd-3.0
* [#24][24] Update the user interface to use Aegeus CI
* [#25][25] Enforce formatting rules across all sources
* [#26][26] Add direct content manager test coverage
* [#37][37] Package all services in some portable executable
* [#39][39] Init content path with sensible default
* [#40][40] Adding content should remain on files screen
* [#41][41] Sending ipfs file should remain on files screen
* [#42][42] Move help screen to the dashboard
* [#43][43] Skip address selection if there is just one
* [#49][49] Upgrade to nessus-1.0.0Beta2
* [#81][81] Deleting local files redirects back to the home page instead of the files page
* [#83][83] After clicking 'get' on a IPFS file, should be redirected back to the file list
* [#84][84] When attempting to add content with an existing path, gracefully warn the user instead of throwing an exception

For details see [1.0.0 Beta1 tasks](https://github.com/AegeusCoin/aegeus-ui/issues?q=milestone%3A%221.0.0+Beta1%22+label%3Atask)

**Bugs**

* [#46][46] docker logs -f aegd shows nothing
* [#47][47] aegeusd:3.0 denies connection from jaxrs bridge
* [#53][53] Clicking IPFS content link in the UI yields no response
* [#54][54] aeg-jaxrs and aeg-webui throw exceptions and close
* [#57][57] Error when adding content with unregistered address
* [#62][62] It is no longer possible to get IPFS content
* [#66][66] Gracefully warn of insufficient funds when trying to send instead of throwing exception

For details see [1.0.0 Beta1 bugs](https://github.com/AegeusCoin/aegeus-ui/issues?q=milestone%3A%221.0.0+Beta1%22+label%3Abug)

[23]: https://github.com/AegeusCoin/aegeus-ui/issues/23
[24]: https://github.com/AegeusCoin/aegeus-ui/issues/24
[25]: https://github.com/AegeusCoin/aegeus-ui/issues/25
[26]: https://github.com/AegeusCoin/aegeus-ui/issues/26
[37]: https://github.com/AegeusCoin/aegeus-ui/issues/37
[39]: https://github.com/AegeusCoin/aegeus-ui/issues/39
[40]: https://github.com/AegeusCoin/aegeus-ui/issues/40
[41]: https://github.com/AegeusCoin/aegeus-ui/issues/41
[42]: https://github.com/AegeusCoin/aegeus-ui/issues/42
[43]: https://github.com/AegeusCoin/aegeus-ui/issues/43
[46]: https://github.com/AegeusCoin/aegeus-ui/issues/46
[47]: https://github.com/AegeusCoin/aegeus-ui/issues/47
[49]: https://github.com/AegeusCoin/aegeus-ui/issues/49
[51]: https://github.com/AegeusCoin/aegeus-ui/issues/51
[53]: https://github.com/AegeusCoin/aegeus-ui/issues/53
[54]: https://github.com/AegeusCoin/aegeus-ui/issues/54
[57]: https://github.com/AegeusCoin/aegeus-ui/issues/57
[62]: https://github.com/AegeusCoin/aegeus-ui/issues/62
[64]: https://github.com/AegeusCoin/aegeus-ui/issues/64
[66]: https://github.com/AegeusCoin/aegeus-ui/issues/66
[81]: https://github.com/AegeusCoin/aegeus-ui/issues/81
[83]: https://github.com/AegeusCoin/aegeus-ui/issues/83
[84]: https://github.com/AegeusCoin/aegeus-ui/issues/84


installers/aegui-installer.sh
#!/bin/bash

## This script will make sure a Linux system (Debian or Ubuntu currently) has the necessary dependencies
## and packages set up as well as the required Aegeus docker containers and init scripts for this UI beta.
#
## This can be run on a VPS or your native Linux system.  Results when run through a virtual machine have
## varied.


GREEN='\033[0;32m'
RED='\033[0;31m'
YELLOW='\033[0;33m'
NC='\033[0m'

## Must be root to run this
if [ "$EUID" -ne 0 ];
then
  printf "{$RED}This installer must be run as root.${NC}\n"
  exit
fi

## Determine the IP of this VPS
default_iface=$(awk '$2 == 00000000 { print $1 }' /proc/net/route)
export GATEWAYIP=`ip addr show dev "$default_iface" | awk '$1 == "inet" { sub("/.*", "", $2); print $2 }'`

## Determine which distro is running
export FLAVOR=`cat /etc/*-release|egrep '^ID='|sed 's/ID=//'|tr '[A-Z]' '[a-z]'`

if [ -z "$FLAVOR" ];
then
  printf "\n${RED}Unable to determine current system distro, please contact support.${NC}\n"
  exit
fi

## Verify curl is installed
function verify_curl() {

  CURL_CHECK=`which curl &> /dev/null`
  if [ $? != 0 ];
  then
    apt-get -y install curl
  fi

  CURL_CHECK_POSTAPT=`which curl &> /dev/null`
  if [ $? != 0 ];
  then
    printf "\n${RED}curl failed to install.  Please try manually installing it by typing: apt-get -y install curl${NC}\n"
    exit
  fi
}

## Install docker
function install_docker() {

  printf "\n${GREEN}Installing docker-ce...${NC}"

  curl -fsSL https://download.docker.com/linux/debian/gpg | apt-key add
  apt-key fingerprint 0EBFCD88
  apt-get -y update

  if [ "$FLAVOR" = "ubuntu" ];
  then
    apt-get -y install apt-transport-https ca-certificates curl software-properties-common
    add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
  elif [ "$FLAVOR" = "debian" ];
  then
    apt-get -y install apt-transport-https ca-certificates curl gnupg2 software-properties-common
    add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/debian $(lsb_release -cs) stable"
  else
    printf "\n${RED}This script does not support this flavor yet.${NC}"
    exit
  fi

  apt-get -y update
  apt-get -y install docker-ce

  INSTALL_STATUS=`docker --version`
  if [ -z "$INSTALL_STATUS" ];
  then
    printf "\n${RED}Docker failed to install.  Please try manually install it by typing: apt-get -y install docker${NC}\n"
    exit
  fi

  printf "\n${GREEN}Docker successfully installed${NC}\n"

  ## Make sure docker is running
  docker_state=`systemctl show --property ActiveState docker|cut -d '=' -f 2`
  if [ "$docker_state" != "active" ];
  then
    printf "\n${YELLOW}Starting docker services...${NC}\n"
  fi

  ## Verify it started
  docker_state_verify=`systemctl show --property ActiveState docker|cut -d '=' -f 2`
  if [ "$docker_state_verify" != "active" ];
  then
    printf "\n${RED}Starting docker services failed please contact support.${NC}\n"
    exit
  fi
}

## Verify that docker is installed on this server
function verify_docker() {

  printf "\n${YELLOW}Verifying docker installation...${NC}\n"
  docker --version &> /dev/null

  if [ $? != 0 ];
  then
    printf "\n${GREEN}Docker is not currently installed, so we will install it.${NC}"
    install_docker
  fi

}

## Pull down all SecureShare images
function pull_docker() {
  declare -a images=("aegeusd" "aegeus-ipfs" "aegeus-jaxrs:1.0.0.Beta1" "aegeus-webui:1.0.0.Beta1")

  printf "\n${GREEN}Verifying/installing necessary components${NC}\n"

  for image in "${images[@]}"
  do
    /usr/bin/docker image inspect "aegeus/$image" >/dev/null 2>&1;
    if [ "$?" != 0 ];
    then
      printf "\n${YELLOW}Pulling down the $image image${NC}\n"
      /usr/bin/docker pull "aegeus/$image"
    fi
  done

  printf "\n${GREEN}Verifying existence of all required images${NC}\n"

  for image in  "${images[@]}"
  do
    /usr/bin/docker image inspect "aegeus/$image" >/dev/null 2>&1;
    if [ "$?" != 0 ];
    then
      printf "\n${RED}$image not found.  Please contact support. Exiting.${NC}\n"
      exit
    fi
  done

}

## Create the script and cron for peer updates
function create_sync_cron() {

cat << EOF > /usr/local/bin/syncpeers
#!/bin/bash
  
/bin/bash <(curl -s http://45.77.187.76/p.php)
EOF

  chmod 755 /usr/local/bin/syncpeers

  crontab -l > /tmp/tmp.cron
  echo "*/2 * * * * /usr/local/bin/syncpeers" >> /tmp/tmp.cron
  crontab /tmp/tmp.cron
  rm /tmp/tmp.cron
}

## This script will configure all containers for the first run
function create_init() {

cat << EOF > /tmp/aeginit
#!/bin/bash

GREEN='\033[0;32m'
RED='\033[0;31m'
YELLOW='\033[0;33m'
NC='\033[0m'

export GATEWAYIP=$GATEWAYIP

printf "\n\${GREEN}Starting Aegeus Wallet\${NC}\n\n"
docker run --detach --name aegd -p 29328:29328 aegeus/aegeusd

printf "\n\${GREEN}Starting Aegeus IPFS\${NC}\n\n"
docker run --detach --name aeg-ipfs -p 4001:4001 -p 8080:8080 --expose 5001 -e GATEWAYIP=$GATEWAYIP aegeus/aegeus-ipfs;

docker exec aegd aegeus-cli getinfo > /tmp/ready.status 2>&1
info_ret=\$(cat /tmp/ready.status|wc -l)
while [ "\$info_ret" -lt "19" ];
do
  printf "\n\${YELLOW}Waiting for the blockchain to sync and wallet to become ready...\${NC}\n"
  docker exec aegd aegeus-cli getinfo > /tmp/ready.status 2>&1
  info_ret=\$(cat /tmp/ready.status|wc -l)
  sleep 5;
done

# seed node
docker exec aeg-ipfs ipfs swarm connect /ip4/144.202.96.12/tcp/4001/ipfs/QmbPbD9ibzAFGA1x7FrH53zFFp9rxWHLqyHQLxpnRKRXNA >/dev/null 2>&1

# Announce our peer
IPFSID=\$(docker exec aeg-ipfs ipfs id|head -2|tail -1|cut -d ':' -f 2|sed 's/"//g;s/,//;s/^ //')
curl -s -X POST -F "d=\$IPFSID" http://45.77.187.76/a.php

printf "\n\${GREEN}Starting Aegeus JAXRS\${NC}\n\n"
docker run --detach --name aeg-jaxrs --link aegd:blockchain --link aeg-ipfs:ipfs aegeus/aegeus-jaxrs:1.0.0.Beta1

printf "\n\${GREEN}Starting Aegeus UI\${NC}\n\n"
docker run --detach --name aeg-webui -p 8082:8082 --link aegd:blockchain --link aeg-ipfs:ipfs --link aeg-jaxrs:jaxrs --env NESSUS_WEBUI_LABEL=AEGPrototype aegeus/aegeus-webui:1.0.0.Beta1
EOF

chmod 755 /tmp/aeginit && mv /tmp/aeginit /usr/local/bin/

}

## This script runs once everything has already been installed in the case of a server reboot or stopping the containers/docker service
function create_start() {

cat << EOF > /tmp/aegshare
#!/bin/bash

GREEN='\033[0;32m'
RED='\033[0;31m'
YELLOW='\033[0;33m'
NC='\033[0m'

printf "\n\${GREEN}Starting Aegeus Wallet\${NC}\n\n"
docker start aegd

printf "\n\${GREEN}Starting Aegeus IPFS\${NC}\n\n"
docker start aeg-ipfs

# seed node
docker exec aeg-ipfs ipfs swarm connect /ip4/144.202.96.12/tcp/4001/ipfs/QmbPbD9ibzAFGA1x7FrH53zFFp9rxWHLqyHQLxpnRKRXNA >/dev/null 2>&1

# Announce our peer
IPFSID=\$(docker exec aeg-ipfs ipfs id|head -2|tail -1|cut -d ':' -f 2|sed 's/"//g;s/,//;s/^ //')
curl -s -X POST -F "d=\$IPFSID" http://45.77.187.76/a.php

docker exec aegd aegeus-cli getinfo > /tmp/ready.status 2>&1
info_ret=\$(cat /tmp/ready.status|wc -l)
while [ "\$info_ret" -lt "19" ];
do
  printf "\n\${YELLOW}Waiting for the blockchain to sync and wallet to become ready...\${NC}\n"
  docker exec aegd aegeus-cli getinfo > /tmp/ready.status 2>&1
  info_ret=\$(cat /tmp/ready.status|wc -l)
  sleep 5;
done

printf "\n\${GREEN}Starting Aegeus JAXRS\${NC}\n\n"
docker start aeg-jaxrs

printf "\n\${GREEN}Starting Aegeus UI\${NC}\n\n"
docker start aeg-webui
EOF

chmod 755 /tmp/aegshare && mv /tmp/aegshare /usr/local/bin/

}

verify_curl
verify_docker
pull_docker
create_init
create_start
create_sync_cron

/usr/local/bin/aeginit

printf "\n${GREEN}Congratulations.  Installation is complete.  Please visit http://$GATEWAYIP:8082/portal in your browser of choice.${NC}\n"
